nemo_megatron_launcher/launch_scripts/nemo_megatron/collections/eval_harness/lm_eval/evaluator.py

```diff
@@ -0,0 +1,165 @@
+import collections
+import itertools
+import logging
+import random
+import time
+
+import lm_eval.metrics
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)-15s | %(name)-7s | %(levelname)-8s: %(message)s")
+logger = logging.getLogger(__name__)
+
+
+def evaluate(
+    lm,
+    task_dict,
+    provide_description,
+    num_fewshot,
+    limit,
+    bootstrap_iters=100000,
+    filter_shots=True,
+    serialize_predictions=False,
+    **kwargs,
+):
+    # TODO: completely refactor this entire function to not be a huge mess, ideally breaking it down into smaller pieces
+    # GEO TODO: I have the impression that a lot of data content is duplicated in many structures
+    #  (task, task_docs, reqs, requests, request_origin). Converting everything to HF dataset objects may be a good alternative
+
+    task_dict_items = [
+        (name, task) for name, task in task_dict.items() if (task.has_validation_docs() or task.has_test_docs())
+    ]
+
+    results = collections.defaultdict(dict)
+    versions = collections.defaultdict(dict)
+
+    requests = collections.defaultdict(list)
+    requests_origin = collections.defaultdict(list)
+
+    # if we ever run into issues where the eval tasks don't fit in memory and we can't afford a machine with bigger memory,
+    # we can always modify this plumbing to support that, but i didn't want to include it just yet because overengineering is bad
+    # (or we could make it write the requests to disk and then read them back out again - probably using an sqlite db because of all the moving parts we have
+
+    # TODO: we need unit tests & sanity checks or something to ensure that the return of `validation_docs` is stable
+
+    docs = {}
+
+    # get lists of each type of request
+    for task_name, task in task_dict_items:
+        versions[task_name] = task.VERSION
+        # default to test doc, fall back to val doc if validation unavailable
+        # TODO: the test-fallback-to-val system isn't final, we should revisit it at some point
+        if task.has_test_docs():
+            task_doc_func = task.test_docs
+        elif task.has_validation_docs():
+            task_doc_func = task.validation_docs
+        else:
+            raise RuntimeError(
+                f"Task {task_name} has neither test docs nor validation docs, please verify data is properly configured for this task."
+            )
+
+        # deterministically shuffle docs and chop off the first `limit` because sometimes docs are in some kind of order
+        task_docs = list(task_doc_func())
+        task_docs = list(
+            zip(range(len(task_docs)), task_docs)
+        )  # use original sample order as ID for evaluation samples
+        rnd = random.Random()
+        rnd.seed(42)
+        rnd.shuffle(task_docs)
+
+        logger.info("Found {} {} documents ...".format(len(task_docs), task_name))
+        logger.info("Building requests for '{}' ...".format(task_name))
```

Comment:
Should  update this to use NeMo logging

Reply:
as @markelsanz14 mentioned, these launcher script is supposed to work independently on a login node without NeMo installation. I am not sure it's a good idea to use NeMo logging in this code. It's supposed to be independent to NeMo src code in the launcher part.